src/repo/remote.rs

```diff
@@ -0,0 +1,692 @@
+        }
+        Ok(())
+    }
+}
+
+/// Loads all remote index (.esi) files from the target directory.
+pub fn load_remotes(base_path: &Path) -> Result<Vec<RemoteIndex>, Error> {
+    let paths = fs::read_dir(base_path).map_err(Error::IOError)?;
+    paths
+        .filter_map(|e| e.ok())
+        .filter(|p| p.path().extension() == Some(OsStr::new(REMOTE_INDEX_EXTENSION)))
+        .map(|p| RemoteIndex::load(p.path()).reify(p.path().display()))
+        .collect()
+}
+
+/// Sends an HTTP GET request with the specified URL.
+///
+/// # Arguments
+/// * `url` - the URL to fetch
+/// * `timeout` - the timeout for the whole of the request
+/// * `size_limit` - the maximum size of the response body, before the
+///   connection is terminated
+/// * `if_modified_since` - Sets the value of the `If-Modified-Since` HTTP
+///   header
+fn open_remote_resource(
+    url: &Url,
+    timeout: Option<Duration>,
+    if_modified_since: Option<SystemTime>,
+) -> Result<Option<(usize, impl Read)>, Error> {
+    let mut agent_builder = AgentBuilder::new();
+    // Alternatively, we could have used Duration::MAX to indicate no timeout.
+    // Unfortunately, the ureq crashes at some unwrap() somewhere then MAX is
+    // provided.
+    if let Some(timeout) = timeout {
+        agent_builder = agent_builder.timeout(timeout);
+    }
+    let agent = agent_builder.build();
+
+    let mut request = agent.get(url.as_ref());
+    if let Some(if_modified_since) = if_modified_since {
+        request = request.set("If-Modified-Since", &format_http_date(if_modified_since));
+    }
+
+    let response = request.call().map_err(|e| Error::HttpError(e.into()))?;
+
+    let status = response.status();
+
+    let content_length = response
+        .header("Content-Length")
+        .unwrap_or("0")
+        .parse::<usize>()
+        .map_err(|_| {
+            Error::HttpError(
+                "The received Content-Length header contained an invalid value!".into(),
+            )
+        })?;
+
+    log::info!(
+        "HTTP GET {} -> {} (Content-Length: {})",
+        url,
+        status,
+        content_length
+    );
+
+    match status {
+        HTTP_STATUS_OK => Ok(Some((content_length, response.into_reader()))),
+        HTTP_STATUS_NOT_MODIFIED => Ok(None),
+        _ => Err(Error::HttpError(
+            format!(
+                "Response status code {} indicates failure!",
+                response.status()
+            )
+            .into(),
+        )),
+    }
+}
+
+fn read_remote_resource(
+    url: &Url,
+    timeout: Duration,
+    size_limit: u64,
+    if_modified_since: Option<SystemTime>,
+) -> Result<Option<Vec<u8>>, Error> {
+    if let Some(reader) = open_remote_resource(url, Some(timeout), if_modified_since)? {
+        let mut body = vec![];
+        reader
+            .1
+            // A malicious third-party could overwhelm might send a very
+            // large response.
+            .take(size_limit)
```

Comment:
FWIW, I don't think we particularly want a limit, or it should be overridable (maybe you already had that in mind). We don't know how big-a-systems people might want to use elfshaker for. Just because something seems inconceivably big to us doesn't mean it isn't useful to someone else.

Reply:
8 MiB which is what the current limit is means elfshaker clone would have to pull 100,000s packs. But you are correct that we don't know how people might use it. How about we make this an environment variable, or maybe remove the limit altogether?

Reply:
I agree that sounds like a lot and I can't imagine anyone doing that in the near future. I'm not sure what this is protecting against in practice -- is it any different than when you run wget/curl, or docker pull?

Thinking aloud, if there were 100k LLVM-sized packs, I guess that would correspond to ~100GiB. But if the packs were 100x smaller it would only be 1GiB, which would be back to the realms of feasible. I could imagine some ways you could achieve this (e.g. if you had one pack per piece of software in nixos, or something like this).